astronomer/providers/snowflake/sensors/snowflake.py

```diff
@@ -74,30 +75,61 @@ def __init__(
-                    dag_id=context["dag"].dag_id,
-                    task_id=context["task"].task_id,
-                    run_id=context["dag_run"].run_id,
-                    snowflake_conn_id=self.snowflake_conn_id,
-                ),
-                method_name=self.execute_complete.__name__,
-            )
+            self._defer(context)
 
-    def execute_complete(self, context: Context, event: dict[str, str] | None = None) -> Any:
+    def _validate_result(self, result: list[tuple[Any]]) -> Any:
+        """Validates query result and verifies if it returns a row"""
+        if not result:
+            if self.fail_on_empty:
+                raise AirflowException("No rows returned, raising as per fail_on_empty flag")
+            else:
+                return False
+
+        first_cell = result[0][0]
+        if self.failure is not None:
+            if callable(self.failure):
+                if self.failure(first_cell):
+                    raise AirflowException(f"Failure criteria met. self.failure({first_cell}) returned True")
+            else:
+                raise AirflowException(f"self.failure is present, but not callable -> {self.failure}")
+        if self.success is not None:
+            if callable(self.success):
+                return self.success(first_cell)
+            else:
+                raise AirflowException(f"self.success is present, but not callable -> {self.success}")
+        return bool(first_cell)
+
+    def _defer(self, context: Context) -> None:
+        self.defer(
+            timeout=timedelta(seconds=self.timeout),
+            trigger=SnowflakeSensorTrigger(
+                sql=self.sql,
+                poke_interval=self.poke_interval,
+                parameters=self.parameters,
+                success=self.success,
+                failure=self.failure,
+                fail_on_empty=self.fail_on_empty,
+                dag_id=context["dag"].dag_id,
+                task_id=context["task"].task_id,
+                run_id=context["dag_run"].run_id,
+                snowflake_conn_id=self.snowflake_conn_id,
+            ),
+            method_name=self.execute_complete.__name__,
+        )
+
+    def execute_complete(self, context: Context, event: dict[str, Any] | None = None) -> Any:
         """
         Callback for when the trigger fires - returns immediately.
         Relies on trigger to throw an exception, otherwise it assumes execution was
         successful.
         """
         if event:
+            if "status" in event and event["status"] == "validate":
+                marker = self._validate_result(event["result"])
+                if marker:
+                    return
+                else:
+                    self._defer(context)
             if "status" in event and event["status"] == "error":
                 raise_error_or_skip_exception(self.soft_fail, event["message"])
             self.log.info(event["message"])
```

Comment:
Not your change, but the else block below looks wrong. When can event be not set? If not set, we should raise an exception?

Reply:
I have fixed this PTAL